crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,434 @@
+        strncmp(level, "WARN", len) == 0 ? OSSL_CMP_LOG_WARNING :
+        strncmp(level, "NOTE", len) == 0 ? OSSL_CMP_LOG_NOTICE :
+        strncmp(level, "INFO", len) == 0 ? OSSL_CMP_LOG_INFO :
+        strncmp(level, "DEBUG", len) == 0 ? OSSL_CMP_LOG_DEBUG :
+        -1;
+}
+
+const char *ossl_cmp_log_parse_metadata(const char *buf,
+                 OSSL_CMP_severity *level, char **func, char **file, int *line)
+{
+    const char *p_func = buf;
+    const char *p_file = buf == NULL ? NULL : strchr(buf, ':');
+    const char *p_level = buf;
+    const char *msg = buf;
+
+    *level = -1;
+    *func = NULL;
+    *file = NULL;
+    *line = 0;
+
+    if (p_file != NULL) {
+        const char *p_line = strchr(++p_file, ':');
+
+        /* check if buf contains at least "CMP "followed by logging level */
+        if ((*level = parse_level(buf)) < 0 && p_line++ != NULL) {
+            /* else check if buf contains location info and logging level */
+            char *p_level_tmp = (char *)p_level;
+            const long line_number = strtol(p_line, &p_level_tmp, 10);
+
+            p_level = p_level_tmp;
+            if (p_level > p_line && *(p_level++) == ':') {
+                if ((*level = parse_level(p_level)) >= 0) {
+                    *func = OPENSSL_strndup(p_func, p_file - 1 - p_func);
+                    *file = OPENSSL_strndup(p_file, p_line - 1 - p_file);
+                    *line = (int)line_number;
+                    msg = strchr(p_level, ':') + 1;
+                    if (*msg == ' ')
+                        msg++;
+                }
+            }
+        }
+    }
+    return msg;
+}
+
+
+/*
+ * auxiliary function for incrementally reporting texts via the error queue
+ */
+static void put_error(int lib, const char *func, int reason,
+                      const char *file, int line)
+{
+    ERR_new();
+    ERR_set_debug(file, line, func);
+    ERR_set_error(lib, reason, NULL);
+}
+
+#define ERR_print_errors_cb_LIMIT 4096 /* size of char buf2[] variable there */
+#define TYPICAL_MAX_OUTPUT_BEFORE_DATA 100
+#define MAX_DATA_LEN (ERR_print_errors_cb_LIMIT-TYPICAL_MAX_OUTPUT_BEFORE_DATA)
+void ossl_cmp_add_error_txt(const char *separator, const char *txt)
+{
+    const char *file;
+    int line;
+    const char *func = NULL;
+    const char *data;
+    int flags;
+    unsigned long err = ERR_peek_last_error();
+
+    if (separator == NULL)
+        separator = "";
+    if (err == 0)
+        put_error(ERR_LIB_CMP, NULL, 0, "", 0);
+
+    do {
+        int available_len;
+        const char *curr = txt, *next = txt;
+        char *tmp;
+
+        ERR_peek_last_error_line_data(&file, &line, &data, &flags);
+        if ((flags & ERR_TXT_STRING) == 0) {
+            data = "";
+            separator = "";
+        }
+        /* TODO add when available: ERR_peek_last_error_func(&func); */
+
+        /* workaround for limit of ERR_print_errors_cb() */
+        available_len = MAX_DATA_LEN - (int)strlen(data) - strlen(separator);
```

Comment:
the type cast here is a bit odd.
why can't this underflow?

Reply:
I've improved the calculation, catching any potential underflows.